### PR TITLE
[OD metrics]: (Step 1/3) Align _BoxRecall to the new object detection API

### PR DESCRIPTION
**Pull request by LukeWood**
# What does this PR do?

Originally, I was planning to migrate COCORecall to `PyMetric`, but I eventually realized that only `MeanAveragePrecision` needs to use `PyMetric`.  Recall can get  numerical accuracy all within graph.  Unfortunately performance is bad, so we will still need to migrate to PyMetric.  

This PR makes Recall a functional metric in the OD API and works with dictionary boxes, but performance is bad.  After this PR:

Outstanding tasks:
- make recall performant
- support metrics in RetinaNet
- support COCOMeanAveragePrecision
- support metrics in MaskRCNN

### CODE REVIEW

**Comment by LukeWood** [comment]
ok @ianstenbit I think this is prepared for a review!

**Comment by LukeWood** [comment]
ok @ianstenbit - the performance of this is quite bad, but I'd like to merge then migrate to PyMetric to improve performance.  PTAL when you have time, some of the utilities + the API will remain after the migration to PyMetric so its good to do 2 PRs.

**Comment by ianstenbit** [comment]
As a reviewer, it's difficult to tell what of these changes are directly related to BoxRecall vs unrelated improvements.

I don't really see the point of this change if we intend to migrate to PyMetric anyway, as this makes a lot of changes to the internals of BoxRecall which is going to change.

Maybe if you separated the utils changes from the BoxRecall changes I would be able to see that more clearly?

**Comment by LukeWood** [comment]
@ianstenbit updated the PR description, and I believe everything in this PR is now related to getting the API of BoxRecall ironed out.  Performance + making the API user facing will be handled in a follow up.

File: keras_cv/metrics/coco/recall.py [hunk 1]
```diff
@@ -22,13 +22,16 @@
...
 
-    A usage guide is available on keras.io:
-    [Using KerasCV COCO metrics](https://keras.io/guides/keras_cv/coco_metrics/).
-    Full implementation details are available in the
-    [KerasCV COCO metrics whitepaper](https://arxiv.org/abs/2207.12120).
+    _BoxRecall is analagous to traditional Recall.  The primary distinction is
+    that when operating in the problem domain of object detection there exists
+    ambiguity in what is considered a true positive.  The _BoxRecall metric
+    works by using the Intersection over Union (IoU) metric to determine whether
+    or not a detection is a true positive or a false positive.  An average
```
Comment (fchollet) [inline keras_cv/metrics/coco/recall.py:32]: Please specify the criteria more explicitly

**Review by fchollet — APPROVED** [review]

File: keras_cv/metrics/coco/recall.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/metrics/coco/recall.py]: Thanks - good point this was ambiguous.  Addressed
